[RHDHBUGS-106][RHIDP-4646] Clarify the behavior of the `NO_PROXY` rules (#709)
redhat-developer/red-hat-developers-documentation-rhdh: modules/admin/procedure-understanding-no-proxy.adoc

[id="understanding-no-proxy"]
= Understanding the `NO_PROXY` exclusion rules

`NO_PROXY` is a comma or space-separated list of hostnames or IP addresses, optionally with port numbers. If the input URL matches any of the entries listed in `NO_PROXY`, then that URL will be fetched by a direct request (i.e., bypassing the proxy settings).

> Note that the default value for `NO_PROXY` in {product-very-short} is `localhost,127.0.0.1`. If you want to override it, please make sure to also include at least `localhost` or `localhost:7007` in the list; otherwise, the {product-very-short} Backend may not work correctly.

Author: ```suggestion
[NOTE]
	====
	The default value for `NO_PROXY` in {product-very-short} is `localhost,127.0.0.1`. If you want to override it, include at least `localhost` or `localhost:7007` in the list. Otherwise, the {product-very-short} Backend may not work correctly.
	====
```

Thanks for the suggestion. I've updated to remove the links to `http://localhost` and `http://localhost:7007`, as we don't need to have links displayed here.

Matching follows the rules below:

* `NO_PROXY=*` will bypass the proxy for all requests.

* Space and commas may be used to separate the entries in the `NO_PROXY` list. For example, `NO_PROXY=localhost,example.com`, `NO_PROXY="localhost example.com"`, or `NO_PROXY="localhost, example.com"` would have the same effect.

* If `NO_PROXY` does not contain any entries, then all requests will be sent through the proxy if the `HTTP(S)_PROXY` settings are configured. Otherwise, requests will be fetched directly.

> * No DNS lookup is performed to decide if a request should bypass the proxy or not. For example, if DNS is known to resolve `example.com` to `1.2.3.4`, setting `NO_PROXY=1.2.3.4` will not have any effect on requests sent to `example.com`. Only requests explicitly sent to the IP address `1.2.3.4` will bypass the proxy.

Author: ```suggestion
* The Backend does not perform a DNS lookup to determine if a request should bypass the proxy or not. For example, if DNS is known to resolve `example.com` to `1.2.3.4`, setting `NO_PROXY=1.2.3.4` will not have any effect on requests sent to `example.com`. Only requests sent to the IP address `1.2.3.4` will bypass the proxy.
```

Just clarifying that the requests originate from the Backend, not the user directly.

> * If a port is added after the hostname or IP Address, then the input request must match both the host/IP and port in order to bypass the proxy. For example, `NO_PROXY=example.com:1234` would exclude requests to `http(s)://example.com:1234` (so calling them directly), but not requests to other ports like `http(s)://example.com` (which will be sent through the proxy).

Author: ```suggestion
* If a port is added after the hostname or IP Address, then the input request must match both the host/IP and port in order to bypass the proxy. For example, `NO_PROXY=example.com:1234` would bypass the proxy for requests to `http(s)://example.com:1234`, but not for requests on other ports, like `http(s)://example.com`.
```

Clarifying the sentence suggested, as I feel like the meaning is changed.

* If no port is specified after the hostname or IP address, all requests to that host/IP address will bypass the proxy regardless of the port. For example, `NO_PROXY=localhost` would exclude all requests sent to `localhost` (so calling them directly), like `http(s)://localhost:7077` and `http(s)://localhost:8888`.

> * IP Address blocks in CIDR notation will not work. So setting `NO_PROXY=10.11.0.0/16` will not have any effect, even if a request is explicitly sent to an IP address in that block.

Author: ```suggestion
* IP Address blocks in CIDR notation will not work. So setting `NO_PROXY=10.11.0.0/16` will not have any effect, even if the Backend sends a request to an IP address in that block.
```

[id="understanding-no-proxy"]
= Understanding the `NO_PROXY` exclusion rules

`NO_PROXY` is a comma or space-separated list of hostnames or IP addresses, optionally with port numbers. If the input URL matches any of the entries listed in `NO_PROXY`, then that URL will be fetched by a direct request (i.e., bypassing the proxy settings).

[NOTE]
====

> 	The default value for `NO_PROXY` in {product-very-short} is `localhost,127.0.0.1`. If you want to override it, include at least `localhost` or `localhost:7007` in the list. Otherwise, the {product-very-short} Backend may not work correctly.

Author: ```suggestion
The default value for `NO_PROXY` in {product-very-short} is `localhost,127.0.0.1`. If you want to override it, include at least `localhost` or `localhost:7007` in the list. Otherwise, the {product-very-short} Backend may not work correctly.
```

> 	The default value for `NO_PROXY` in {product-very-short} is `localhost,127.0.0.1`. If you want to override it, include at least `localhost` or `localhost:7007` in the list. Otherwise, the {product-very-short} Backend may not work correctly.

Author: nvm, I think it was because of the space in the beginning.

> 	The default value for `NO_PROXY` in {product-very-short} is `localhost,127.0.0.1`. If you want to override it, include at least `localhost` or `localhost:7007` in the list. Otherwise, the {product-very-short} Backend may not work correctly.

Reviewer: @rm3l , I see that removing the spaces from the note solved the earlier problem. I am reviewing the changes again for confirmation.

====

Matching follows the rules below:

* `NO_PROXY=*` will bypass the proxy for all requests.

> * Space and commas might separate the entries in the `NO_PROXY` list. For example, `NO_PROXY=localhost,example.com`, `NO_PROXY="localhost example.com"`, or `NO_PROXY="localhost, example.com"` would have the same effect.

Author: ```suggestion
* Space and commas might separate the entries in the `NO_PROXY` list. For example, `NO_PROXY="localhost,example.com"`, or `NO_PROXY="localhost example.com"`, or `NO_PROXY="localhost, example.com"` would have the same effect.
```

Protecting the example env var value to separate it from the rest of the sentence. Also consistent with how the other examples are rendered.

> * You may use space and commas may to separate the entries in the `NO_PROXY` list. For example, `NO_PROXY=localhost,example.com`, `NO_PROXY="localhost example.com"`, or `NO_PROXY="localhost, example.com"` would have the same effect.

Author: ```suggestion
* You may use space and commas to separate the entries in the `NO_PROXY` list. For example, `NO_PROXY=localhost,example.com`, `NO_PROXY="localhost example.com"`, or `NO_PROXY="localhost, example.com"` would have the same effect.
```

> * If `NO_PROXY` contains no entries, configuring the `HTTP(S)_PROXY` settings will make the Backend send all requests through the proxy. Otherwise, it fetches requests directly.

Author: ```suggestion
* If `NO_PROXY` contains no entries, configuring the `HTTP(S)_PROXY` settings will make the Backend send all requests through the proxy.
```

* The Backend does not perform a DNS lookup to determine if a request should bypass the proxy or not. For example, if DNS is known to resolve `example.com` to `1.2.3.4`, setting `NO_PROXY=1.2.3.4` will not have any effect on requests sent to `example.com`. Only requests sent to the IP address `1.2.3.4` will bypass the proxy.

* If a port is added after the hostname or IP Address, then the input request must match both the host/IP and port in order to bypass the proxy. For example, `NO_PROXY=example.com:1234` would bypass the proxy for requests to `http(s)://example.com:1234`, but not for requests on other ports, like `http(s)://example.com`.

> * If no port is specified after the hostname or IP address, all requests to that host/IP address will bypass the proxy regardless of the port. For example, `NO_PROXY=localhost` would bypass the proxy for requests sent to URLs like `http(s)://localhost:7077` and `http(s)://localhost:8888`.

Author: `if you fail to specify a port [...]` makes it sound like setting the port is mandatory IMO, which is not the case here. How about just stating `If you do not specify a port [...]` ?

> * If no port is specified after the hostname or IP address, all requests to that host/IP address will bypass the proxy regardless of the port. For example, `NO_PROXY=localhost` would bypass the proxy for requests sent to URLs like `http(s)://localhost:7077` and `http(s)://localhost:8888`.

Reviewer: Sure, ` If you do not specify a port [...] ` looks good here. You can use that instead.

> * If no port is specified after the hostname or IP address, all requests to that host/IP address will bypass the proxy regardless of the port. For example, `NO_PROXY=localhost` would bypass the proxy for requests sent to URLs like `http(s)://localhost:7077` and `http(s)://localhost:8888`.

Author: ```suggestion
* If you do not specify a port after the hostname or IP address, all requests to that host/IP address will bypass the proxy regardless of the port. For example, `NO_PROXY=localhost` would bypass the proxy for requests sent to URLs like `http(s)://localhost:7077` and `http(s)://localhost:8888`.
```

> * If no port is specified after the hostname or IP address, all requests to that host/IP address will bypass the proxy regardless of the port. For example, `NO_PROXY=localhost` would exclude all requests sent to `localhost` (so calling them directly), like `http(s)://localhost:7077` and `http(s)://localhost:8888`.

Author: ```suggestion
* If no port is specified after the hostname or IP address, all requests to that host/IP address will bypass the proxy regardless of the port. For example, `NO_PROXY=localhost` would bypass the proxy for requests sent to URLs like `http(s)://localhost:7077` and `http(s)://localhost:8888`.
```
